R/preprocess.R

```diff
@@ -1,37 +1,69 @@
+# TODO: Add mc.cores to parallelise across samples?
 preprocessRaw <- function(rgSet) {
     .isRGOrStop(rgSet)
+    if (.isHDF5ArrayBacked(rgSet)) {
+        BACKEND <- "HDF5Array"
+    } else {
+        BACKEND <- NULL
+    }
     locusNames <- getManifestInfo(rgSet, "locusNames")
-    M <- matrix(NA_real_, ncol = ncol(rgSet), nrow = length(locusNames),
-                dimnames = list(locusNames, colnames(rgSet)))
-    U <- M
-    TypeII.Name <- getProbeInfo(rgSet, type = "II")$Name
-    M[TypeII.Name,] <- getGreen(rgSet)[getProbeInfo(rgSet, type = "II")$AddressA,]
-    U[TypeII.Name,] <- getRed(rgSet)[getProbeInfo(rgSet, type = "II")$AddressA,]
+    # HDF5: Idea is to process sample-by-sample (column-by-column). Each
+    #       iteration loads the red and green data into memory as a matrix,
+    #       then subsets  by rows, then writes back to disk. This is an
+    #       efficient strategy because each sample's data is contiguous in the
+    #       HDF5 file. If instead we first did rowwise subsetting, then
+    #       reading the data from the HDF5 file incurs a large overhead due to
+    #       non-contiguous reads.
+    TypeII <- getProbeInfo(rgSet, type = "II")
     TypeI.Red <- getProbeInfo(rgSet, type = "I-Red")
     TypeI.Green <- getProbeInfo(rgSet, type = "I-Green")
-    M[TypeI.Red$Name,] <- getRed(rgSet)[TypeI.Red$AddressB,]
-    M[TypeI.Green$Name,] <- getGreen(rgSet)[TypeI.Green$AddressB,]
-    U[TypeI.Red$Name,] <- getRed(rgSet)[TypeI.Red$AddressA,]
-    U[TypeI.Green$Name,] <- getGreen(rgSet)[TypeI.Green$AddressA,]
-    out <- MethylSet(Meth = M, Unmeth = U, colData = colData(rgSet),
-                     annotation = annotation(rgSet), metadata = metadata(rgSet))
+    M_and_U <- lapply(sampleNames(rgSet), function(this_sample) {
```

Comment:
This is prototypical of a reasonably efficient way to work with "long" _DelayedMatrix_ when you need them to be atomic R matrix/vector: work on a column at a time, `realize()` the result using the appropriate/requested backend, then cbind() the results. Furthermore, the columns can be done in parallel